lib/explicit_activerecord/transaction_guard.rb

```diff
@@ -0,0 +1,74 @@
+# typed: true
+
+# prevent callers from calling your methods within a transaction,
+# providing you with confidence that you can perform non-atomic operations, or mutate records without something downstream triggering a rollback,
+#
+# class MyService
+#   include ExplicitActiveRecord::TransactionGuard
+#
+#   def my_non_atomic_method(args)
+#     ensure_transaction_integrity!
+#     SidekiqWorker.perform_async(args)
+#   end
+#
+#   private
+#
+#   def trust_me_im_not_in_a_transaction(args)
+#     allow_transaction do
+#       my_non_atomic_method(args)
+#     end
+#   end
+# end
+
+module ExplicitActiveRecord
+  module TransactionGuard
+    extend T::Sig
+
+    class ForbiddenCallWhileInDbTransaction < StandardError; end
+
+    class << self
+      extend T::Sig
+
+      sig { returns(T.nilable(Integer)) }
+      attr_accessor :allowed_transaction_count
+
+      sig { returns(T.nilable(T::Boolean)) }
+      attr_accessor :ignore
+
+      sig { returns(Integer) }
+      def open_transactions
+        ActiveRecord::Base.connection.open_transactions
+      end
+
+      sig { returns(Integer) }
+      def excessive_transactions
+        open_transactions - (allowed_transaction_count || 0)
+      end
+    end
+
+    sig { void }
+    def ensure_transaction_integrity!
+      unless TransactionGuard.ignore == true || TransactionGuard.excessive_transactions <= 0
+        Kernel.raise ForbiddenCallWhileInDbTransaction, "This method guards against open transactions. You have called it with #{TransactionGuard.excessive_transactions} open tranactions."
+      end
+    end
+
+    sig { params(blk: T.proc.returns(T.untyped)).returns(T.untyped) }
+    def ignore_transaction_integrity!(&blk)
+      TransactionGuard.ignore = true
+      yield
+    ensure
+      TransactionGuard.ignore = false
+    end
+
+    sig { params(blk: T.proc.returns(T.untyped)).returns(T.untyped) }
+    def allow_transaction(&blk)
+      TransactionGuard.allowed_transaction_count ||= 0
+      TransactionGuard.allowed_transaction_count = T.must(TransactionGuard.allowed_transaction_count) + 1
```

Comment:
So the idea is that allow transaction works in a nested way?

Reply:
in short, yeah. in test, all of HI runs in a transaction (except for request specs), so we can't just think of it as binary in a transaction or not in a transaction. that's why this uses counts of open transactions rather than `open_transactions > 0` (or the private-to-ARBase `in_transaction?`, which does the same thing).

there are two concerns i have with this:

1. counting things and assigning them to a static variable seems very likely to be no good, in cases where we fail to decrease the count, or threading. it's for this reason that i'm not 100% sure even _i_ would approve this PR. (i'm hoping to get more confidence by running some tests using this diff in HI)
2. `open_transactions` only increments if you use `joinable: false, requires_new: true`, which the rails team says over and over again to _not_ use. in other words, i'm not sure which problem is bigger -- that we are nesting transactions and thus making non-atomic operations risky, or that we're using less-than-good API that prevents us from using `after_commit` hooks.